components/crypto_dot_com/browser/crypto_dot_com_service.cc

```diff
@@ -152,40 +171,245 @@ bool CryptoDotComService::GetAssetRankings(
+    GetDepositAddressCallback callback,
+    const std::string& asset,
+    const int status,
+    const std::string& body,
+    const std::map<std::string, std::string>& headers) {
+  DVLOG(2) << __func__ << ": " << body;
+  auto value = CryptoDotComJSONParser::GetValidDepositAddress(body);
+  if (value.is_none()) {
+    auto empty_deposit = base::JSONReader::Read(kEmptyDepositAddress);
+    return std::move(callback).Run(std::move(*empty_deposit));
+  }
+
+  std::move(callback).Run(std::move(value));
+}
+
+bool CryptoDotComService::GetNewsEvents(GetNewsEventsCallback callback) {
+  auto internal_callback =
+      base::BindOnce(&CryptoDotComService::OnGetNewsEvents,
+                     base::Unretained(this), std::move(callback));
+  net::HttpRequestHeaders headers;
+  headers.SetHeader("widget-token", access_token_);
+  GURL url(kCryptoDotComGetNewsEventsURL);
+  return NetworkRequest(url, "GET", "", headers, std::move(internal_callback));
+}
+
+void CryptoDotComService::OnGetNewsEvents(
+    GetNewsEventsCallback callback,
+    const int status,
+    const std::string& body,
+    const std::map<std::string, std::string>& headers) {
+  DVLOG(2) << __func__ << ": " << body;
+  auto value = CryptoDotComJSONParser::GetValidNewsEvents(body);
+  if (value.is_none()) {
+    auto empty_news_events = base::JSONReader::Read(kEmptyNewsEvents);
+    return std::move(callback).Run(
+        empty_news_events->FindListKey("events")->Clone());
+  }
+
+  std::move(callback).Run(std::move(value));
+}
+
+bool CryptoDotComService::CreateMarketOrder(
+    base::Value order,
+    CreateMarketOrderCallback callback) {
+  auto internal_callback =
+      base::BindOnce(&CryptoDotComService::OnCreateMarketOrder,
+                     base::Unretained(this), std::move(callback));
+  net::HttpRequestHeaders headers;
+  headers.SetHeader("widget-token", access_token_);
+  GURL url(kCryptoDotComCreateMarketOrderURL);
+
+  std::string body;
+  base::JSONWriter::Write(order, &body);
+  DVLOG(2) << __func__ << ": " << body;
+  return NetworkRequest(url, "POST", body, headers,
+                        std::move(internal_callback));
+}
+
+void CryptoDotComService::OnCreateMarketOrder(
+    CreateMarketOrderCallback callback,
+    const int status,
+    const std::string& body,
+    const std::map<std::string, std::string>& headers) {
+  DVLOG(2) << __func__ << ": " << body;
+  std::move(callback).Run(CryptoDotComJSONParser::GetValidOrderResult(body));
+}
+
+std::string CryptoDotComService::GetAuthClientUrl() const {
+  return kCryptoDotComAuthURL;
+}
+
+bool CryptoDotComService::SetAccessToken(const std::string& access_token) {
+  access_token_ = access_token;
+
+  PrefService* prefs = user_prefs::UserPrefs::Get(context_);
+  if (access_token_.empty()) {
+    prefs->SetString(kCryptoDotComAccessToken, access_token_);
+    return true;
+  }
+
+  std::string encrypted_access_token;
+  if (!OSCrypt::EncryptString(access_token, &encrypted_access_token)) {
```

Comment:
It is worth noting that `OSCrypt::EncryptString()` unfortunately reuses the same IV in CBC mode for the cipher. This is a chromium concern (and not necessarily a bug - there are usecases where this is needed, hence why it is there). What does this mean? If we reuse the same access_token for any reason a malicious party will be able to distinguish. In the context of this pull request this shouldn't be concerning at all (AFAIU), but the use of `OSCrypt::EncryptString()` should be carefully thought through in our other uses of it in the future. This use case is obviously fine, but at least it is noted and can be looked out for in the future. 

@diracdeltas I will add this to our security audit check list to make sure this is carefully considered if there ever was to be a privacy violation due to misuse of this method (leaking data due to encrypting the same message twice).

Reply:
@orspetol great find. could you please open a follow up issue to audit for unsafe uses of oscrypt? we also use it for sync encryption and IIRC on-disk encryption of private rewards data. cc @darkdh @zenparsing @tmancey

Reply:
https://github.com/brave/brave-browser/issues/15127

Reply:
@diracdeltas I can confirm ads does not use OSCrypt at this time. Thanks